### PR TITLE
Improvement of error reporting

### PR DESCRIPTION
**Pull request by thomaseizinger**
This PR brings a few improvements to the way we use our error reporting library. Most notable, instead of relying on the global error handler to be called before showing the dialog, we now use an (undocumented) event of the library itself to always show the report dialog after an event was logged to Sentry.

In addition to this change, I also changed the way our config is overridden per environment as this made it easier to locally test these changes.

For more information, read the commit messages.

### CODE REVIEW

File: src/store/attachStoreStateFactory.js [hunk 1]
```diff
@@ -0,0 +1,6 @@
+export default function(store) {
+  return function attachStoreState(data, existingCallback = (data) => data) {
```
Comment (duffleit) [inline src/store/attachStoreStateFactory.js]: paramtername `existingCallback`: Raven calls this `original`. So probably `originalCallback` is a better name.

File: src/main.js [hunk 2]
```diff
@@ -22,12 +23,16 @@ const browserHistory = useRouterHistory(createBrowserHistory)({
...
-const store = createStore(initialState, browserHistory)
+const initialState = window.___INITIAL_STATE__;
+const store = createStore(initialState, browserHistory);
 const history = syncHistoryWithStore(browserHistory, store, {
   selectLocationState: (state) => state.router
 });
 
+// Configure Raven to always attach the current state of the store to the event
+let attachStoreState = attachStoreStateFactory(store);
+Raven.setDataCallback(attachStoreState);
```
Comment (duffleit) [inline src/main.js]: As already mentioned, this api-function is not documented. So we have no safety and can not rely on it to be available in future versions. Seems to be the only way to get things done, so it is okay.

Comment (thomaseizinger) [reply src/main.js]: The callback interface is documented and will probably not break without a new major version. The only thing that is not documented is the ravenSuccess event that is used for always triggering the feedback dialog after an error.
This functionality here, including our callback implementation is perfectly fine use of Raven's public API.

If your comment targets the setter, this is also public API in order to change the config after initialization.

File: src/store/attachStoreStateFactory.js (hunk 1 above)
Comment (thomaseizinger) [reply src/store/attachStoreStateFactory.js]: Makes sense, I will change that.

**Comment by thomaseizinger** [comment]
I just thought of something:

Initially, we used the `window.onerror` handler to show the report dialog. This allowed some situations in which an error was triggered and reported to Sentry but no report dialog was shown. This can happen if an error is triggered but not bubbled up to the `window.onerror` handler (which is the last thing in the chain).

Yesterday, I had a situation in which the app worked just fine but still produced an error (which was logged in Sentry). In this case, it would have been a bad idea to show the report dialog to the user because for them, there was no error. (I don't yet know, what exactly happened.)

With the changes of this PR concerning the report dialog, we always show the report dialog after the `ravenSuccess` event, which is triggered as soon as Raven successfully reports an issue to Sentry. This would have been the case yesterday.

What are your thoughts on this @duffleit ?

**Comment by thomaseizinger** [comment]
I decided to merge the current implementation into the master branch because the scenario described above is rather rare. The only reason why the user did not "see" anything to be wrong was because the failure happened during the update of the HTML title attribute.
Although there might be scenarios in which the user can still use the app, something does not work as expected. Therefore, I think it is ok to always show the dialog to the user because everything else would be a lot of work to configure which is not worth it IMO in the current situation.

**Pull request merged**